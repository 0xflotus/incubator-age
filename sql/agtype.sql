--
-- AGTYPE data type regression tests
--

--
-- Load extension and set path
--
LOAD 'agensgraph';
SET search_path TO ag_catalog;

--
-- Create a table using the AGTYPE type
--
CREATE TABLE agtype_table (type text, agtype agtype);

--
-- Insert values to exercise agtype_in/agtype_out
--
INSERT INTO agtype_table VALUES ('bool', 'true');
INSERT INTO agtype_table VALUES ('bool', 'false');

INSERT INTO agtype_table VALUES ('null', 'null');

INSERT INTO agtype_table VALUES ('string', '""');
INSERT INTO agtype_table VALUES ('string', '"This is a string"');

INSERT INTO agtype_table VALUES ('integer', '0');
INSERT INTO agtype_table VALUES ('integer', '9223372036854775807');
INSERT INTO agtype_table VALUES ('integer', '-9223372036854775808');

INSERT INTO agtype_table VALUES ('float', '0.0');
INSERT INTO agtype_table VALUES ('float', '1.0');
INSERT INTO agtype_table VALUES ('float', '-1.0');
INSERT INTO agtype_table VALUES ('float', '100000000.000001');
INSERT INTO agtype_table VALUES ('float', '-100000000.000001');
INSERT INTO agtype_table VALUES ('float', '0.00000000000000012345');
INSERT INTO agtype_table VALUES ('float', '-0.00000000000000012345');

INSERT INTO agtype_table VALUES ('integer array',
	'[-9223372036854775808, -1, 0, 1, 9223372036854775807]');
INSERT INTO agtype_table VALUES('float array',
	'[-0.00000000000000012345, -100000000.000001, -1.0, 0.0, 1.0, 100000000.000001, 0.00000000000000012345]');
INSERT INTO agtype_table VALUES('mixed array', '[true, false, null, "string", 1, 1.0, {"bool":true}]');

INSERT INTO agtype_table VALUES('object', '{"bool":true, "null":null, "string":"string", "integer":1, "float":1.2, "arrayi":[-1,0,1], "arrayf":[-1.0, 0.0, 1.0], "object":{"bool":true, "null":null, "string":"string", "int":1, "float":8.0}}');

--
-- Special float values: NaN, +/- Infinity
--
INSERT INTO agtype_table VALUES ('float  nan', 'nan');
INSERT INTO agtype_table VALUES ('float  Infinity', 'Infinity');
INSERT INTO agtype_table VALUES ('float -Infinity', '-Infinity');
INSERT INTO agtype_table VALUES ('float  inf', 'inf');
INSERT INTO agtype_table VALUES ('float -inf', '-inf');

SELECT * FROM agtype_table;

--
-- These should fail
--
INSERT INTO agtype_table VALUES ('bad integer', '9223372036854775808');
INSERT INTO agtype_table VALUES ('bad integer', '-9223372036854775809');
INSERT INTO agtype_table VALUES ('bad float', '-NaN');
INSERT INTO agtype_table VALUES ('bad float', 'Infi');
INSERT INTO agtype_table VALUES ('bad float', '-Infi');

--
-- Test agtype mathematical operator functions
-- +, -, unary -, *, /, %, and ^
--
SELECT agtype_add('1', '-1');
SELECT agtype_add('1', '-1.0');
SELECT agtype_add('1.0', '-1');
SELECT agtype_add('1.0', '-1.0');

SELECT agtype_sub('-1', '-1');
SELECT agtype_sub('-1', '-1.0');
SELECT agtype_sub('-1.0', '-1');
SELECT agtype_sub('-1.0', '-1.0');

SELECT agtype_neg('-1');
SELECT agtype_neg('-1.0');
SELECT agtype_neg('0');
SELECT agtype_neg('0.0');

SELECT agtype_mul('-2', '3');
SELECT agtype_mul('2', '-3.0');
SELECT agtype_mul('-2.0', '3');
SELECT agtype_mul('2.0', '-3.0');

SELECT agtype_div('-4', '3');
SELECT agtype_div('4', '-3.0');
SELECT agtype_div('-4.0', '3');
SELECT agtype_div('4.0', '-3.0');

SELECT agtype_pow('-2', '3');
SELECT agtype_pow('2', '-1.0');
SELECT agtype_pow('2.0', '3');
SELECT agtype_pow('2.0', '-1.0');

--
-- Should fail with divide by zero
--
SELECT agtype_div('1', '0');
SELECT agtype_div('1', '0.0');
SELECT agtype_div('1.0', '0');
SELECT agtype_div('1.0', '0.0');

--
-- Should get Infinity
--
SELECT agtype_pow('0', '-1');
SELECT agtype_pow('-0.0', '-1');

--
-- Test operators +, -, unary -, *, /, %, and ^
--
SELECT '3.14'::agtype + '3.14'::agtype;
SELECT '3.14'::agtype - '3.14'::agtype;
SELECT -'3.14'::agtype;
SELECT '3.14'::agtype * '3.14'::agtype;
SELECT '3.14'::agtype / '3.14'::agtype;
SELECT '3.14'::agtype % '3.14'::agtype;
SELECT '3.14'::agtype ^ '2'::agtype;

--
-- Test orderability of comparison operators =, <>, <, >, <=, >=
-- These should all return true
-- Integer
SELECT agtype_in('1') = agtype_in('1');
SELECT agtype_in('1') <> agtype_in('2');
SELECT agtype_in('1') <> agtype_in('-2');
SELECT agtype_in('1') < agtype_in('2');
SELECT agtype_in('1') > agtype_in('-2');
SELECT agtype_in('1') <= agtype_in('2');
SELECT agtype_in('1') >= agtype_in('-2');

-- Float
SELECT agtype_in('1.01') = agtype_in('1.01');
SELECT agtype_in('1.01') <> agtype_in('1.001');
SELECT agtype_in('1.01') <> agtype_in('1.011');
SELECT agtype_in('1.01') < agtype_in('1.011');
SELECT agtype_in('1.01') > agtype_in('1.001');
SELECT agtype_in('1.01') <= agtype_in('1.011');
SELECT agtype_in('1.01') >= agtype_in('1.001');
SELECT agtype_in('1.01') < agtype_in('Infinity');
SELECT agtype_in('1.01') > agtype_in('-Infinity');
-- NaN, under ordering, is considered to be the biggest numeric value
-- greater than positive infinity. So, greater than any other number.
SELECT agtype_in('1.01') < agtype_in('NaN');
SELECT agtype_in('NaN') > agtype_in('Infinity');
SELECT agtype_in('NaN') > agtype_in('-Infinity');
SELECT agtype_in('NaN') = agtype_in('NaN');

-- Mixed Integer and Float
SELECT agtype_in('1') = agtype_in('1.0');
SELECT agtype_in('1') <> agtype_in('1.001');
SELECT agtype_in('1') <> agtype_in('0.999999');
SELECT agtype_in('1') < agtype_in('1.001');
SELECT agtype_in('1') > agtype_in('0.999999');
SELECT agtype_in('1') <= agtype_in('1.001');
SELECT agtype_in('1') >= agtype_in('0.999999');
SELECT agtype_in('1') < agtype_in('Infinity');
SELECT agtype_in('1') > agtype_in('-Infinity');
SELECT agtype_in('1') < agtype_in('NaN');

-- Mixed Float and Integer
SELECT agtype_in('1.0') = agtype_in('1');
SELECT agtype_in('1.001') <> agtype_in('1');
SELECT agtype_in('0.999999') <> agtype_in('1');
SELECT agtype_in('1.001') > agtype_in('1');
SELECT agtype_in('0.999999') < agtype_in('1');

-- Strings
SELECT agtype_in('"a"') = agtype_in('"a"');
SELECT agtype_in('"a"') <> agtype_in('"b"');
SELECT agtype_in('"a"') < agtype_in('"aa"');
SELECT agtype_in('"b"') > agtype_in('"aa"');
SELECT agtype_in('"a"') <= agtype_in('"aa"');
SELECT agtype_in('"b"') >= agtype_in('"aa"');

-- Lists
SELECT agtype_in('[0, 1, null, 2]') = agtype_in('[0, 1, null, 2]');
SELECT agtype_in('[0, 1, null, 2]') <> agtype_in('[2, null, 1, 0]');
SELECT agtype_in('[0, 1, null]') < agtype_in('[0, 1, null, 2]');
SELECT agtype_in('[1, 1, null, 2]') > agtype_in('[0, 1, null, 2]');

-- Objects (Maps)
SELECT agtype_in('{"bool":true, "null": null}') = agtype_in('{"null":null, "bool":true}');
SELECT agtype_in('{"bool":true}') < agtype_in('{"bool":true, "null": null}');

-- Comparisons between types
-- Object < List < String < Boolean < Integer = Float = Numeric < Null
SELECT agtype_in('1') < agtype_in('null');
SELECT agtype_in('NaN') < agtype_in('null');
SELECT agtype_in('Infinity') < agtype_in('null');
SELECT agtype_in('true') < agtype_in('1');
SELECT agtype_in('true') < agtype_in('NaN');
SELECT agtype_in('true') < agtype_in('Infinity');
SELECT agtype_in('"string"') < agtype_in('true');
SELECT agtype_in('[1,3,5,7,9,11]') < agtype_in('"string"');
SELECT agtype_in('{"bool":true, "integer":1}') < agtype_in('[1,3,5,7,9,11]');
SELECT agtype_in('[1, "string"]') < agtype_in('[1, 1]');
SELECT agtype_in('{"bool":true, "integer":1}') < agtype_in('{"bool":true, "integer":null}');

--
-- Cleanup
--
DROP TABLE agtype_table;

--
-- End of AGTYPE data type regression tests
--
