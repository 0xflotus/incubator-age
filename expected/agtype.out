--
-- AGTYPE data type regression tests
--
--
-- Load extension and set path
--
LOAD 'agensgraph';
SET search_path TO ag_catalog;
--
-- Create a table using the AGTYPE type
--
CREATE TABLE agtype_table (type text, agtype agtype);
--
-- Insert values to exercise agtype_in/agtype_out
--
INSERT INTO agtype_table VALUES ('bool', 'true');
INSERT INTO agtype_table VALUES ('bool', 'false');
INSERT INTO agtype_table VALUES ('null', 'null');
INSERT INTO agtype_table VALUES ('string', '""');
INSERT INTO agtype_table VALUES ('string', '"This is a string"');
INSERT INTO agtype_table VALUES ('integer', '0');
INSERT INTO agtype_table VALUES ('integer', '9223372036854775807');
INSERT INTO agtype_table VALUES ('integer', '-9223372036854775808');
INSERT INTO agtype_table VALUES ('float', '0.0');
INSERT INTO agtype_table VALUES ('float', '1.0');
INSERT INTO agtype_table VALUES ('float', '-1.0');
INSERT INTO agtype_table VALUES ('float', '100000000.000001');
INSERT INTO agtype_table VALUES ('float', '-100000000.000001');
INSERT INTO agtype_table VALUES ('float', '0.00000000000000012345');
INSERT INTO agtype_table VALUES ('float', '-0.00000000000000012345');
INSERT INTO agtype_table VALUES ('integer array',
	'[-9223372036854775808, -1, 0, 1, 9223372036854775807]');
INSERT INTO agtype_table VALUES('float array',
	'[-0.00000000000000012345, -100000000.000001, -1.0, 0.0, 1.0, 100000000.000001, 0.00000000000000012345]');
INSERT INTO agtype_table VALUES('mixed array', '[true, false, null, "string", 1, 1.0, {"bool":true}]');
INSERT INTO agtype_table VALUES('object', '{"bool":true, "null":null, "string":"string", "integer":1, "float":1.2, "arrayi":[-1,0,1], "arrayf":[-1.0, 0.0, 1.0], "object":{"bool":true, "null":null, "string":"string", "int":1, "float":8.0}}');
--
-- Special float values: NaN, +/- Infinity
--
INSERT INTO agtype_table VALUES ('float  nan', 'nan');
INSERT INTO agtype_table VALUES ('float  Infinity', 'Infinity');
INSERT INTO agtype_table VALUES ('float -Infinity', '-Infinity');
INSERT INTO agtype_table VALUES ('float  inf', 'inf');
INSERT INTO agtype_table VALUES ('float -inf', '-inf');
SELECT * FROM agtype_table;
      type       |                                                                                                       agtype                                                                                                       
-----------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 bool            | true
 bool            | false
 null            | null
 string          | ""
 string          | "This is a string"
 integer         | 0
 integer         | 9223372036854775807
 integer         | -9223372036854775808
 float           | 0.0
 float           | 1.0
 float           | -1.0
 float           | 100000000.000001
 float           | -100000000.000001
 float           | 1.2345e-16
 float           | -1.2345e-16
 integer array   | [-9223372036854775808, -1, 0, 1, 9223372036854775807]
 float array     | [-1.2345e-16, -100000000.000001, -1.0, 0.0, 1.0, 100000000.000001, 1.2345e-16]
 mixed array     | [true, false, null, "string", 1, 1.0, {"bool": true}]
 object          | {"bool": true, "null": null, "float": 1.2, "arrayf": [-1.0, 0.0, 1.0], "arrayi": [-1, 0, 1], "object": {"int": 1, "bool": true, "null": null, "float": 8.0, "string": "string"}, "string": "string", "integer": 1}
 float  nan      | NaN
 float  Infinity | Infinity
 float -Infinity | -Infinity
 float  inf      | Infinity
 float -inf      | -Infinity
(24 rows)

--
-- These should fail
--
INSERT INTO agtype_table VALUES ('bad integer', '9223372036854775808');
ERROR:  value "9223372036854775808" is out of range for type bigint
LINE 1: INSERT INTO agtype_table VALUES ('bad integer', '92233720368...
                                                        ^
INSERT INTO agtype_table VALUES ('bad integer', '-9223372036854775809');
ERROR:  value "-9223372036854775809" is out of range for type bigint
LINE 1: INSERT INTO agtype_table VALUES ('bad integer', '-9223372036...
                                                        ^
INSERT INTO agtype_table VALUES ('bad float', '-NaN');
ERROR:  invalid input syntax for type agtype
LINE 1: INSERT INTO agtype_table VALUES ('bad float', '-NaN');
                                                      ^
DETAIL:  Token "-NaN" is invalid.
CONTEXT:  AGTYPE data, line 1: -NaN
INSERT INTO agtype_table VALUES ('bad float', 'Infi');
ERROR:  invalid input syntax for type agtype
LINE 1: INSERT INTO agtype_table VALUES ('bad float', 'Infi');
                                                      ^
DETAIL:  Token "Infi" is invalid.
CONTEXT:  AGTYPE data, line 1: Infi
INSERT INTO agtype_table VALUES ('bad float', '-Infi');
ERROR:  invalid input syntax for type agtype
LINE 1: INSERT INTO agtype_table VALUES ('bad float', '-Infi');
                                                      ^
DETAIL:  Token "-Infi" is invalid.
CONTEXT:  AGTYPE data, line 1: -Infi
--
-- Test agtype mathematical operator functions
-- +, -, unary -, *, /, %, and ^
--
SELECT agtype_add('1', '-1');
 agtype_add 
------------
 0
(1 row)

SELECT agtype_add('1', '-1.0');
 agtype_add 
------------
 0.0
(1 row)

SELECT agtype_add('1.0', '-1');
 agtype_add 
------------
 0.0
(1 row)

SELECT agtype_add('1.0', '-1.0');
 agtype_add 
------------
 0.0
(1 row)

SELECT agtype_sub('-1', '-1');
 agtype_sub 
------------
 0
(1 row)

SELECT agtype_sub('-1', '-1.0');
 agtype_sub 
------------
 0.0
(1 row)

SELECT agtype_sub('-1.0', '-1');
 agtype_sub 
------------
 0.0
(1 row)

SELECT agtype_sub('-1.0', '-1.0');
 agtype_sub 
------------
 0.0
(1 row)

SELECT agtype_neg('-1');
 agtype_neg 
------------
 1
(1 row)

SELECT agtype_neg('-1.0');
 agtype_neg 
------------
 1.0
(1 row)

SELECT agtype_neg('0');
 agtype_neg 
------------
 0
(1 row)

SELECT agtype_neg('0.0');
 agtype_neg 
------------
 -0.0
(1 row)

SELECT agtype_mul('-2', '3');
 agtype_mul 
------------
 -6
(1 row)

SELECT agtype_mul('2', '-3.0');
 agtype_mul 
------------
 -6.0
(1 row)

SELECT agtype_mul('-2.0', '3');
 agtype_mul 
------------
 -6.0
(1 row)

SELECT agtype_mul('2.0', '-3.0');
 agtype_mul 
------------
 -6.0
(1 row)

SELECT agtype_div('-4', '3');
 agtype_div 
------------
 -1
(1 row)

SELECT agtype_div('4', '-3.0');
    agtype_div     
-------------------
 -1.33333333333333
(1 row)

SELECT agtype_div('-4.0', '3');
    agtype_div     
-------------------
 -1.33333333333333
(1 row)

SELECT agtype_div('4.0', '-3.0');
    agtype_div     
-------------------
 -1.33333333333333
(1 row)

SELECT agtype_pow('-2', '3');
 agtype_pow 
------------
 -8.0
(1 row)

SELECT agtype_pow('2', '-1.0');
 agtype_pow 
------------
 0.5
(1 row)

SELECT agtype_pow('2.0', '3');
 agtype_pow 
------------
 8.0
(1 row)

SELECT agtype_pow('2.0', '-1.0');
 agtype_pow 
------------
 0.5
(1 row)

--
-- Should faile with divide by zero
--
SELECT agtype_div('1', '0');
ERROR:  division by zero
SELECT agtype_div('1', '0.0');
ERROR:  division by zero
SELECT agtype_div('1.0', '0');
ERROR:  division by zero
SELECT agtype_div('1.0', '0.0');
ERROR:  division by zero
--
-- Should get Infinity
--
SELECT agtype_pow('0', '-1');
 agtype_pow 
------------
 Infinity
(1 row)

SELECT agtype_pow('-0.0', '-1');
 agtype_pow 
------------
 -Infinity
(1 row)

--
-- Test operators +, -, unary -, *, /, %, and ^
--
SELECT '3.14'::agtype + '3.14'::agtype;
 ?column? 
----------
 6.28
(1 row)

SELECT '3.14'::agtype - '3.14'::agtype;
 ?column? 
----------
 0.0
(1 row)

SELECT -'3.14'::agtype;
 ?column? 
----------
 -3.14
(1 row)

SELECT '3.14'::agtype * '3.14'::agtype;
 ?column? 
----------
 9.8596
(1 row)

SELECT '3.14'::agtype / '3.14'::agtype;
 ?column? 
----------
 1.0
(1 row)

SELECT '3.14'::agtype % '3.14'::agtype;
 ?column? 
----------
 0.0
(1 row)

SELECT '3.14'::agtype ^ '2'::agtype;
 ?column? 
----------
 9.8596
(1 row)

--
-- Cleanup
--
DROP TABLE agtype_table;
--
-- End of AGTYPE data type regression tests
--
